File: LeetCode/Medium/Jump Game/Java/jumpGame.java
// recursive solution dp approach
class Solution {
    public boolean canJump(int[] nums) {
        return recur(nums, 1);
    }

    private boolean recur(int[] nums, int ind) {
        if (ind >= nums.length) return true;
        if (nums[ind-1] > 0) {
            if (nums[ind-1] > nums[ind]) nums[ind] = nums[ind-1]-1;
            return recur(nums, ind+1);
        }
        return false;
    }
}

//greedy approach
class Solution {
    public boolean canJump(int[] nums) {
        int jump = nums[0];
        for (int i=1; i<nums.length; i++) {
            if (jump <= 0) return false; 
            jump = Math.max(jump-1, nums[i]);
        }
        return jump >= 0;
    }
}

File: LeetCode/Medium/Jump Game/Go/jump_game.go
package main

import "fmt"

func jump_game(nums []int) bool {
	if nums[0] == 0 {
		return len(nums) == 1
	}
	jump := nums[0]
	for i := 1; i < len(nums); i++ {
		if jump == 0 {
			return false
		}
		jump = max(jump-1, nums[i])
	}
	return true
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(jump_game([]int{2, 3, 1, 1, 4}))
}
